core/slim_sim.cpp

```diff
@@ -6449,22 +6466,35 @@ void SLiMSim::WritePopulationTable(tsk_table_collection_t *p_tables)
 			migration_index++;
 		}
 		
-		tsk_id_t tsk_population = tsk_population_table_add_row(&p_tables->populations, (char *)metadata_rec, (uint32_t)metadata_length);
+		tsk_population_id = tsk_population_table_add_row(&p_tables->populations, (char *)metadata_rec, (uint32_t)metadata_length);
 		last_id_written++;
+		assert(tsk_population_id == last_id_written);
 		
 		free(metadata_rec);
 		
-		if (tsk_population < 0) handle_error("tsk_population_table_add_row", tsk_population);
+		if (tsk_population_id < 0) handle_error("tsk_population_table_add_row", tsk_population_id);
 	}
 	
 	// finally, write out empty entries for the rest of the table; empty entries are needed
 	// up to largest_subpop_id_ because there could be ancestral nodes that reference them
-	while (last_id_written < last_subpop_id)
+	while (last_id_written < (slim_objectid_t) last_subpop_id)
 	{
-		tsk_id_t tsk_population = tsk_population_table_add_row(&p_tables->populations, (char *)&last_subpop_id, (uint32_t)0);	// the address is unused
-		if (tsk_population < 0) handle_error("tsk_population_table_add_row", tsk_population);
 		last_id_written++;
+		if (last_id_written < (slim_objectid_t) population_table_copy->num_rows) {
+			tsk_population_t tsk_population_object;
+			tsk_population_table_get_row(population_table_copy, last_id_written, &tsk_population_object);
+			if (ret != 0) handle_error("WritePopulationTable  tsk_population_table_get_row()", ret);
+			tsk_population_id = tsk_population_table_add_row(&p_tables->populations, tsk_population_object.metadata, tsk_population_object.metadata_length);
+			if (tsk_population_id < 0) handle_error("tsk_population_table_add_row", tsk_population_id);
+		} else {
+			tsk_population_id = tsk_population_table_add_row(
+					&p_tables->populations,
+					NULL, 0);
+			if (tsk_population_id < 0) handle_error("tsk_population_table_add_row", tsk_population_id);
+		}
+		assert(tsk_population_id == last_id_written);
 	}
+	free(population_table_copy);
```

Comment:
isn't there a tskit free-the-internals-of-this function we're supposed to call before free() here?